crates/ruff_python_parser/src/parser/helpers.rs

```diff
@@ -47,11 +48,52 @@ pub(super) const fn token_kind_to_cmp_op(tokens: [TokenKind; 2]) -> Option<CmpOp
 /// Helper for `parse_decorators` to determine if `expr` is a [`dotted_name`] from the decorator
 /// grammar before Python 3.9.
 ///
+/// Returns `None` if `expr` is a `dotted_name`. Returns `Some((description, range))` if it is not,
+/// where `description` is a string describing the invalid node and `range` is the node's range.
+///
 /// [`dotted_name`]: https://docs.python.org/3.8/reference/compound_stmts.html#grammar-token-dotted-name
-pub(super) fn is_name_or_attribute_expression(expr: &Expr) -> bool {
-    match expr {
-        Expr::Attribute(attr) => is_name_or_attribute_expression(&attr.value),
-        Expr::Name(_) => true,
-        _ => false,
-    }
+pub(super) fn invalid_pre_py39_decorator_node(expr: &Expr) -> Option<(&'static str, TextRange)> {
```

Comment:
Nit: The function name suggests that it returns a node where, in fact, it returns a message and a range. 

I think we should either change the method to return a `Option<&Expr>` with a second method that allows deriving a display name for the expression or rename the method